### ISSUE TITLE
Make optimizer available for access in engine

### ISSUE THREAD

**Issue by akandadai** [pytorch/ignite#737]
I think it would be nice to have access to the optimizer through the ignite.engine.Engine class for more customization.
For example if I wanted to tune momentum while training I could do the following:
```python
@trainer.on(ignite.engine.Events.ITERATION_COMPLETED)
def change_momentum(engine):
   engine.optimizer.momentum *= 1.001
```
Currently this is not possible because optimizer is local to the `_update()` function used to initialize `ignite.engine.Engine`

This would also open the door to supporting logging of gradients and weights during training I think.

**Comment by vfdev-5** [comment]
@akandadai this is possible by setting the optimizer to the state like that:

```python
import torch
model = torch.nn.Linear(1, 1)
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.1)

from ignite.engine import Engine, Events

trainer = Engine(lambda e, b: None)

@trainer.on(Events.STARTED)
def update_momentum(engine):
    engine.state.optimizer = optimizer


@trainer.on(Events.ITERATION_STARTED, optimizer)
def update_momentum(engine, opt):
    engine.state.optimizer.param_groups[0]['momentum'] += 0.01

trainer.run([0, 1, 2, 3, 4, 5])
optimizer.param_groups[0]['momentum']
> 0.16000000000000003
```
or simply passing optimizer as a param to the handler:
```python
import torch
model = torch.nn.Linear(1, 1)
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.1)

from ignite.engine import Engine, Events

trainer = Engine(lambda e, b: None)

@trainer.on(Events.ITERATION_STARTED, optimizer)
def update_momentum(engine, opt):
    opt.param_groups[0]['momentum'] += 0.01

trainer.run([0, 1, 2, 3, 4, 5])
optimizer.param_groups[0]['momentum']
> 0.16000000000000003
```

**Comment by vfdev-5** [comment]
> This would also open the door to supporting logging of gradients and weights during training I think.

This is already done with tensorboard/visdom/etc loggers: 
- https://pytorch.org/ignite/contrib/handlers.html#ignite.contrib.handlers.tensorboard_logger.WeightsScalarHandler
- https://pytorch.org/ignite/contrib/handlers.html#ignite.contrib.handlers.tensorboard_logger.GradsScalarHandler

**Comment by akandadai** [comment]
Awesome! You guys were really thorough
